### ISSUE TITLE
Laws and tests for Divariant

### ISSUE THREAD

**Issue by lemastero** [zio/zio-prelude#290]
- [ ] define laws using predicates
- [ ] add instances for Divariant
- [ ] add support for law testing for Divariant (possibly needs change in zio-test)
- [ ] add tests for instances using ZSpec

I am able to add laws using predicates like in `IdentityCompose`
```scala
  def identityCompose[A, B](
    ab: A :=> B
  )(implicit eq: Equal[A :=> B]): Boolean = {
    val ab1 = compose(identity[B], ab)
    val ab2 = compose(ab, identity[A])

    eq.equal(ab1, ab) && eq.equal(ab2, ab)
  }
```
and instances, as I have done a decent amount of work around profunctors (Divariant).
To handle laws using Lawful/Laws/.... I would need some hints on how to approach this.

**Comment by sideeffffect** [comment]
I actually tried that, but I found the Lawful/Laws/... from ZIO Test insufficient :disappointed: 
We would need in ZIO Test something like this
```scala
package object laws {
  object LawfulF {
    type Divariant[-CapsF[_[-_, +_]], -Caps[_]] = ZLawfulF.Divariant[CapsF, Caps, Any]
  }
  object LawsF {
    type Divariant[-CapsF[_[-_, +_]], -Caps[_]] = ZLawsF.Divariant[CapsF, Caps, Any]
```

Given that, then we would need to write `Derive2` and `Derive2Equal` and `coherent.DivariantDerive2Equal` but that should already be simple enough.

**Comment by sideeffffect** [comment]
ticket in ZIO to implement this: https://github.com/zio/zio/issues/4211

**Comment by adamgfraser** [comment]
Yes, the downside of the laws framework is that you need a different instance for each kind since we don't really have a way to abstract over kindedness in Scala 2, but that is definitely something we can add!
